metadata-ingestion/src/datahub/ingestion/source/schema/json_schema.py

```diff
@@ -0,0 +1,364 @@
+    description="Extracts schemas, following references",
+)
+@capability(
+    capability_name=SourceCapability.PLATFORM_INSTANCE,
+    description="Supports platform instance via config",
+)
+@capability(
+    capability_name=SourceCapability.DESCRIPTIONS,
+    description="Extracts descriptions at top level and field level",
+)
+@capability(
+    capability_name=SourceCapability.DELETION_DETECTION,
+    description="With stateful ingestion enabled, will remove entities from DataHub if they are no longer present in the source",
+)
+@capability(
+    capability_name=SourceCapability.TAGS,
+    description="Does not currently support extracting tags",
+    supported=False,
+)
+@capability(
+    capability_name=SourceCapability.OWNERSHIP,
+    description="Does not currently support extracting ownership",
+    supported=False,
+)
+@dataclass
+class JsonSchemaSource(StatefulIngestionSourceBase):
+    """
+    This source extracts metadata from a single JSON Schema or multiple JSON Schemas rooted at a particular path.
+    It performs reference resolution based on the `$ref` keyword.
+
+    Metadata mapping:
+    - Schemas are mapped to Datasets with sub-type Schema
+    - The name of the Schema (Dataset) is inferred from the `$id` property and if that is missing, the file name.
+    - Browse paths are minted based on the path
+    """
+
+    @staticmethod
+    def _pull_up_reference(schema_dict):
+        """Transforms a schema that is of the form
+        {
+         "$ref": "#definitions/MyDefinition",
+         "definitions": {
+            "MyDefinition": {
+                ... definition
+            },
+            "OtherDefinition": {
+                ... other definition
+            }
+         }
+        }
+        to
+        {
+            ... definition
+
+            "definitions": {
+            "OtherDefinition": {
+                ... other definition
+            }
+         }
+        """
+
+        if "$ref" in schema_dict:
+            reference = schema_dict["$ref"]
+            if reference.startswith("#/"):
+                parts = reference[2:].split("/")
+                referred = schema_dict
+                for p in parts:
+                    referred = referred[p]
+                for k, v in referred.items():
+                    schema_dict[k] = v
+                deletable = schema_dict
+                for p in parts[:-1]:
+                    deletable = deletable[p]
+                deletable.pop(parts[-1])
+                schema_dict.pop("$ref")
+                if "title" not in schema_dict:
+                    schema_dict["title"] = parts[-1]
+        return schema_dict
+
+    @staticmethod
+    def _load_json_schema(filename, loader, use_id_as_base_uri):
+        """Loads the given schema file"""
+        path = Path(filename).resolve()
+        base_path = dirname(str(path))
+        base_uri = "file://{}/".format(base_path)
+
+        with open(path) as schema_file:
```

Comment:
Thoughts on wrapping this open in try:except

Reply:
outer try catch per file should be enough here